Log superuser project entry

dannyroberts opened the pull request:
This just logs superuser access (every 6 hours per project). What do you think?

dannyroberts committed:
log superuser entry to a domain every 6 hours

dannyroberts committed:
add proactive index to SuperuserProjectEntryRecord

since it'll be queried every new superuser project entry

emord commented on corehq/apps/domain/project_access/models.py:
```diff
@@ -0,0 +1,26 @@
+ENTRY_RECORD_FREQUENCY = datetime.timedelta(hours=6)
+
+
+class SuperuserProjectEntryRecord(models.Model):
```
This is probably going to end up being a long list as HQ grows. should we partition this table now? (maybe by week or month?) http://architect.readthedocs.io/features/partition/postgresql.html

emord commented on corehq/apps/domain/project_access/models.py:
```diff
@@ -0,0 +1,26 @@
+    def entry_recently_recorded(cls, username, domain):
+        return cls.objects.filter(
+            username=username,
+            project=domain,
```
Should we be more consistent on project/domain here?

snopoke commented on corehq/apps/domain/project_access/middleware.py:
```diff
@@ -0,0 +1,18 @@
+                and hasattr(request, 'domain'):
+            return self.record_entry(request.domain, request.couch_user.username)
+
+    @atomic
```
Will this not start a transaction even if it's a cache hit? Might be better to put `atomic` as the first decorator.

snopoke commented on corehq/apps/domain/project_access/models.py:
```diff
@@ -0,0 +1,26 @@
+            username=username,
+            project=domain,
+            last_login__gt=datetime.datetime.utcnow() - ENTRY_RECORD_FREQUENCY,
+        ).count() > 0
```
It's more efficient to do a `.exists()` query

emord commented on corehq/apps/domain/project_access/models.py:
```diff
@@ -0,0 +1,26 @@
+
+
+class SuperuserProjectEntryRecord(models.Model):
+    username = models.EmailField()
```
I think this should have an index on it. It may be useful to look at all of someone's actions if their account is compromised. or maybe even a proactive email that says "this is what you did today"

dannyroberts committed:
tweak SuperuserProjectEntryRecord

- partition by month
- add index
- use exists rather than count in a query
- re-make migrations

dannyroberts committed:
add architect partition to SuperuserProjectEntryRecord migration

dannyroberts committed:
put quickcache around atomic

dannyroberts committed:
remove unnecessary @atomic decorator

dannyroberts commented:
Going ahead and merging, having judged that the potential for performance issues due to the extra db checks and writes will be minimal, since they are only done after checking that `request.couch_user.is_superuser`.

emord commented:
>  judged that the potential for performance issues due to the extra db checks and writes will be minimal, since they are only done after checking that request.couch_user.is_superuser.

no change necessary, but it seems that logging like this could be done asynchronously through kafka/celery in the future (there's probably a lot more places that we could follow that practice. don't htink it's a big deal here)
